### ISSUE TITLE
Cloud control support (and firmware 2.7)

### ISSUE THREAD

**Issue by AlexxIT** [AlexxIT/SonoffLAN#99]
Hi all. I completely rewrote a component to support management both local and cloud control at the same time.

Why I added cloud support?

- devices with 2nd firmware do not support the latest API of LAN control
- so many users have problems setting up multicast traffic
- some users have problems with multiple routers
- some users have devices in other locations
- for some users, the router freezes from multicast traffic
- after rebooting HA, devices do not always appear immediately (`unavailable` problem)
- devices may sometimes not respond to a local command (`can't send switch to` problem)

If you want me to add support for your device: share it through the **eWeLink** mobile application to my account sonofflan@gmail.com

Waiting for feedback from you.

**Comment by AdmiralStipe** [comment]
For now it looks great, I restarted HA a couple of times and both my Sonoff devices are available and responding immediately after boot. 
Also the Log problem described in issue [#97](https://github.com/AlexxIT/SonoffLAN/issues/97) is gone... well, not really gone but it's not tied to Sonoff anymore, now it reads
```
2020-05-11 14:14:55 WARNING (zeroconf-ServiceBrowser__axis-video._tcp.local.) [zeroconf] Exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/zeroconf/__init__.py", line 2644, in send
    bytes_sent = s.sendto(packet, 0, (real_addr, port))
OSError: [Errno 126] No error information
```
So it might have not been directly related to Sonoff after all... 
Will keep monitoring and report if troubles occur.

**Comment by maciak** [comment]
Hi,
My only device on FW2.7 got discovered properly. It's The LED strip one.
But after I try to change the state it does nothing and connection from connected changes to error.
`
manufacturer ABC
model 2.0-led
sw version PSF-BTA-GL v2.7.0
connection error
rssi -56
`

**Comment by alesoft73** [comment]
Thanks for your Work.
The Sonoff r2 POW attributes are not updated after switch go to off.

**Comment by AlexxIT** [comment]
@maciak @alesoft73 you can write me in Discord or Telegram.
And share devices to my acc. I do not have them for debugging.

**Comment by maciak** [comment]
@AlexxIT Sure, I just don't know what discord server you are on. Don't know your telegram either ;]

**Comment by AlexxIT** [comment]
@maciak added support for Sonoff LED
![download-1](https://user-images.githubusercontent.com/511909/81646115-58116780-9433-11ea-8307-fb33e3b15bb7.png)

**Comment by AlexxIT** [comment]
And fixed shared devices control

**Comment by AlexxIT** [comment]
The component cannot work simultaneously with the 3rd version mobile application.
You need either an application of the 4th version (Android only).
Or create a second account, share devices with it and use it password in the component.

**Comment by maciak** [comment]
@AlexxIT Works great! Only missing is sync to music but that's actually shit. Not usable at all. it has a mic and on any sound, it randomly changes colors.

**Comment by maciak** [comment]
I found one other issue with L1 LED. I have it on a schedule in ewelink app. It does not update the status in HA when light is turned on or off by it.

**Comment by peterbuga** [comment]
hi @AlexxIT great work you've done here, glad to see you took it to another level 👍 ! 

small notice, i'm not sure if you're aware / been contacted already or if it has mentioned already (i quickly read through issues titles, even closed ones) but itead/coolkit contacted me to let me know that the current appid+secret **won't work anymore after Aug 31st 2020** (they were also kind enough to provide me new ones for personal use only)

so, either take in account this change and expect the parts depending on it to fail in the near future or start looking into the newer app logic [cough] (personally i never got the time to do it, not sure if i'll be able anymore). 
either way i'm interested on your take on this and please drop a mention of my username if you chose to pursue to continue having the cloud support and maybe i can do some help too.

cheers 🍻!

**Comment by AlexxIT** [comment]
@peterbuga they also sent a key to me. But the key only works with sonoff devices. He does not see third-party manufacturers. Now I solve this issue with ewelink...

**Comment by peterbuga** [comment]
> @peterbuga they also sent a appid+secret to me. But the appid only works with sonoff devices. It does not see third-party manufacturers. Now I solve this issue with itead...

well that's a bummer, i'm gonna follow the development and see where this goes.

> The link to your component has always been in my readme :)

i know and i highly appreciate that ✊ !

**Comment by AlexxIT** [comment]
> I found one other issue with L1 LED. I have it on a schedule in ewelink app. It does not update the status in HA when light is turned on or off by it.

I checked this case. Everything is working. Maybe there were problems connecting to ewelink servers.

**Comment by AlexxIT** [comment]
> Only missing is sync to music but that's actually shit.

Fixed music mode.

And cloud temperature for TH16.

**Comment by AlexxIT** [comment]
I need feedback from **iFan2** and **iFan3**.
For test **only local** and **only cloud** mode.

**Comment by ashrafmn** [comment]
Hi Alex,, 
great work thanks,, do you support the sonoff SC?
if not i will share mine with you for future support

thanks gain,
Ashraf

**Comment by AlexxIT** [comment]
@ashrafmn please, share.

**Comment by ashrafmn** [comment]
@AlexxIT couldn't share the device, it says User 'sonofflan@gmail.com' not found in ewelink app


the SC device is appearing as cloud, with all attributes below is the logs:

2020-05-15 15:32:12 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 15:32:17 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 15:37:06 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'rssi': -36, 'fwVersion': '2.7.0', 'staMac': 'BC:DD:XX:XX:XX:XX'}
2020-05-15 15:37:22 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 15:37:28 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 15:38:06 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'rssi': -31, 'fwVersion': '2.7.0', 'staMac': 'BC:DD:XX:XX:XX:XX'}
2020-05-15 15:39:06 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'rssi': -34, 'fwVersion': '2.7.0', 'staMac': 'BC:DD:XX:XX:XX:XX'}
2020-05-15 15:52:16 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 15:52:21 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 15:52:28 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 15:52:32 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:05:06 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:05:11 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:27:34 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:27:37 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:33:40 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:33:44 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:44:56 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:44:59 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:45:07 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:45:12 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:50:23 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:50:27 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 16:50:30 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 16:50:35 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 17:00:48 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'online': False}
2020-05-15 17:02:39 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'online': True}
2020-05-15 17:02:40 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'sledOnline': 'on'}
2020-05-15 17:02:40 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'rssi': -27, 'fwVersion': '2.7.0', 'staMac': 'BC:DD:XX:XX:XX:XX'}
2020-05-15 17:02:43 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 17:02:56 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 17:03:59 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'rssi': -34, 'fwVersion': '2.7.0', 'staMac': 'BC:DD:XX:XX:XX:XX'}
2020-05-15 17:09:22 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 17:09:25 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 17:12:31 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 17:12:35 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 17:12:51 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 17:12:54 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}
2020-05-15 17:21:54 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 6, 'dusty': 2}
2020-05-15 17:21:59 DEBUG (MainThread) [custom_components.sonoff.sonoff_cloud] Cloud3 <= id: 1000xxxxxx, seq: None | {'humidity': 15, 'temperature': 23, 'light': 10, 'noise': 5, 'dusty': 2}






![Select more devices](https://user-images.githubusercontent.com/48861077/82062593-025aea80-96d3-11ea-8a1c-59a45079f824.jpg)

**Comment by AlexxIT** [comment]
@ashrafmn do you see such a sensor?

Other users today also have a problem with sharing. Yesterday everything worked.

**Comment by ashrafmn** [comment]
@AlexxIT  yes it appeared as binary sensor,, plz refer to screenshot from my state:

and by the way,, the user sharing to this account : "sonofflan@gmail.com" is not working.

![Screen Shot 2020-05-15 at 9 40 00 PM](https://user-images.githubusercontent.com/48861077/82085205-4959d780-96f5-11ea-8369-925ea8a18054.png)

**Comment by AlexxIT** [comment]
@ashrafmn try share again. It's easier to me have device in the list. I have shared today on this account.

**Comment by ashrafmn** [comment]
@AlexxIT still sharing to sonofflan@gmail.com is not working "User not found"

**Comment by federom** [comment]
Hi @AlexxIT , 

Thanks for your work on this ! I was willing to include the B1 (2.6 fw) in HA for a long time !

I was trying to install this beta version, but in the integration installation page i just can select the version, not the branch (master) as described in the .md. Is there something else i am missing ?

Thanks!

**Comment by AlexxIT** [comment]
@federom try scroll down in dropdown list. This is the last item

**Comment by federom** [comment]
Thanks @AlexxIT !

**Comment by AlexxIT** [comment]
@ashrafmn try to share again, maybe fixed. If it doesn’t work out - write me in Telegram or Discord (AlexxIT#0816)

**Comment by panagiotis81** [comment]
Great work. Besides the B1 which works great, it actually integrated my EACHEN WIFI SMART DOOR/WINDOW SENSOR-WDW too!!!

**Comment by AlexxIT** [comment]
@panagiotis81 EACHEN WIFI SMART SENSOR works like a sensor? Displays on/off?

**Comment by panagiotis81** [comment]
> @panagiotis81 EACHEN WIFI SMART SENSOR works like a sensor? Displays on/off?

@AlexxIT It finds it as a switch with on/off state, I created a sensor for it with in configuration and work fine. I could share it for you if you want.
 
Entity
switch.sonoff_1000_dxxxxx
 
State
off
State attributes (YAML, optional)

manufacturer: coolkit
model: OPL-DMA
sw_version: OPL-DMA-GL v1000.2.917
connection: cloud
rssi: -62
friendly_name: Door sensor
supported_features: 0

1
manufacturer: coolkit
2
model: OPL-DMA
3
sw_version: OPL-DMA-GL v1000.2.917
4
connection: cloud
5
rssi: -62
6
friendly_name: Door sensor
7
supported_features: 0
8
​

**Comment by federom** [comment]
It works for me with a B1 !! Sometimes i need to press more than once in order to make it work but it works !!

Thanks @AlexxIT for your work !

**Comment by AlexxIT** [comment]
@panagiotis81 can you share it to my acc to check if all works fine? sonofflan@gmail.com

**Comment by panagiotis81** [comment]
@AlexxIT Done!

**Comment by ashrafmn** [comment]
@AlexxIT sharing still not working with me
What country should I use?


![E2884ECC-8C96-487A-AFDC-081E78324B95](https://user-images.githubusercontent.com/48861077/82475076-9dd5cc00-9ad4-11ea-9f61-a5829d050ccf.png)

**Comment by AlexxIT** [comment]
@panagiotis81 supported Door/Window Sensor in latest master (and binary `device_class`)

@ashrafmn supported Sonoff SC in latest master

**Comment by ashrafmn** [comment]
@AlexxIT yes confirmed,, i updated to latest master component and all sensors were loaded as sensors,,, thanks Alex for your amazing work :)

plz keep it up and running,, a lot of automations are now depending on your work

**Comment by panagiotis81** [comment]
@AlexxIT I checked too, deleted my binary sensor, updated to latest, and only a binary sensor is visible now without the switch, so nobody has to create it manually. Great work as always.

**Comment by AlexxIT** [comment]
Hey. I have new BIG update for you.
This is **unavailable** state for local and cloud mode.
It was difficult, but I hope everything works well.

If something doesn’t work well, I have a new debug mode for you:
https://youtu.be/Lt5fT4N5Pm8

**Comment by mattsaxon** [comment]
@AlexxIT this works great, I'm in the process of migrating my configuration over to use your component in place of mine and I'll update you with any issues I find. Great work!

**Comment by maciak** [comment]
After this update I lost control over my L1 LED.  Anything I can do? It's unavailable all the time.

**Comment by AlexxIT** [comment]
@maciak update once again. Everything will return to its place.

**Comment by maciak** [comment]
Sorry, might have been false alarm as I've noticed HACS now updates to new version not master. Does anyone know how to choose master now in the new gui?

**Comment by AlexxIT** [comment]
I hope release 2.0 is near.

**Comment by GutoFagundes007** [comment]
I can't find the integration of sonoff in hacks anymore. Can someone help me?

**Comment by federom** [comment]
Same here ! It seems that after the update i loss the b1

**Comment by AlexxIT** [comment]
@federom check which version you updated to.

**Comment by djgragra** [comment]
Hi I used your AlexxIT / SonoffLAN repository and it works well, the only device that I can't add in the Home Assistant are the Sonoff L1 LEDs, is there any way to use them inside HA?

**Comment by AlexxIT** [comment]
@djgragra L1 supported in latest master (beta) version

**Comment by federom** [comment]
@AlexxIT it prompts me to update to 1.4.. there is no drop-down to select the version

**Comment by AlexxIT** [comment]
@federom problem with new HACS version. You can select master from: SonoffLAN > 3 dots > reinstall > drop-down last item

**Comment by AlexxIT** [comment]
Version 2.0 released. You no longer need to use the maser version.

**Comment by Niconath** [comment]
> Sortie de la version 2.0. Vous n'avez plus besoin d'utiliser la version maser.

Hi, it works with HA 0.110.2 and HACS v 1.0.3.
I already need to remove the custom_component sonoff from peterbuga before installing yours?Thanks

**Comment by djgragra** [comment]
> @djgragra L1 supported in latest master (beta) version

Thanks, where can I find the master (beta) and the installation instructions?

**Comment by AlexxIT** [comment]
@Niconath @djgragra From now, it is supported in the latest stable version:
https://github.com/AlexxIT/SonoffLAN#install-with-hacs

**Comment by Niconath** [comment]
> @Niconath @djgragra A partir de maintenant, il est pris en charge dans la dernière version stable:
> https://github.com/AlexxIT/SonoffLAN#install-with-hacs

I have installed, everything works but in HACS I have the impression that it appears not to be installed. It's weird, isn't it? Thanks to you!
![Capture d’écran (12)](https://user-images.githubusercontent.com/63063404/82894684-419afe00-9f53-11ea-931d-f0affd52f887.png)

**Comment by djgragra** [comment]
> @djgragra L1 supported in latest master (beta) version

It's works great! Good job!

**Comment by federom** [comment]
Does anyone use it with Hue Integration in Node Red ? I am being unable to get the components now 

2020-05-26 12:26:22 ERROR (MainThread) [homeassistant.setup] Error during setup of component sonoff
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 171, in _async_setup_component
    hass, processed_config
  File "/config/custom_components/sonoff/__init__.py", line 192, in async_setup
    cameras = EWeLinkCameras()
  File "/config/custom_components/sonoff/sonoff_camera.py", line 78, in __init__
    self.sock.bind(('', 50000))
OSError: [Errno 98] Address in use

**Comment by AlexxIT** [comment]
@federom fixed in v2.0.2

**Comment by federom** [comment]
It works now ! Thanks !
